materialize-motherduck/driver.go

```diff
@@ -0,0 +1,474 @@
+			return fmt.Errorf("updating spec: %w", err)
+		}
+		return nil
+	})
+}
+
+func (c client) InstallFence(ctx context.Context, checkpoints sql.Table, fence sql.Fence) (sql.Fence, error) {
+	var err = c.withDB(func(db *stdsql.DB) error {
+		var err error
+		fence, err = sql.StdInstallFence(ctx, db, checkpoints, fence, base64.StdEncoding.DecodeString)
+		return err
+	})
+	return fence, err
+}
+
+func (c client) withDB(fn func(*stdsql.DB) error) error {
+	return fn(c.db)
+}
+
+type transactor struct {
+	cfg *config
+
+	fence     sql.Fence
+	storeConn *stdsql.Conn
+
+	bindings []*binding
+}
+
+func newTransactor(
+	ctx context.Context,
+	ep *sql.Endpoint,
+	fence sql.Fence,
+	bindings []sql.Table,
+) (_ pm.Transactor, err error) {
+	cfg := ep.Config.(*config)
+
+	s3client, err := cfg.toS3Client(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	storeConn, err := ep.Client.(client).db.Conn(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("creating store connection: %w", err)
+	}
+
+	t := &transactor{
+		cfg:       cfg,
+		storeConn: storeConn,
+		fence:     fence,
+	}
+
+	for _, b := range bindings {
+		t.bindings = append(t.bindings, &binding{
+			target:    b,
+			storeFile: newStagedFile(s3client, cfg.Bucket, cfg.BucketPath, b.Columns()),
+		})
+	}
+
+	return t, nil
+}
+
+type binding struct {
+	target    sql.Table
+	storeFile *stagedFile
+}
+
+func (d *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage) error) error {
+	for it.Next() {
+		panic("connector must be set to delta updates")
+	}
+	return nil
+}
+
+func (d *transactor) Store(it *pm.StoreIterator) (pm.StartCommitFunc, error) {
+	ctx := it.Context()
+
+	for it.Next() {
+		b := d.bindings[it.Binding]
+		b.storeFile.start()
+
+		if converted, err := b.target.ConvertAll(it.Key, it.Values, it.RawJSON); err != nil {
+			return nil, fmt.Errorf("converting store parameters: %w", err)
+		} else if err := b.storeFile.encodeRow(ctx, converted); err != nil {
+			return nil, fmt.Errorf("encoding row for store: %w", err)
+		}
+	}
+	if it.Err() != nil {
+		return nil, it.Err()
+	}
+
+	return func(ctx context.Context, runtimeCheckpoint *protocol.Checkpoint, _ <-chan struct{}) (*pf.ConnectorState, pf.OpFuture) {
+		var err error
+		if d.fence.Checkpoint, err = runtimeCheckpoint.Marshal(); err != nil {
+			return nil, pf.FinishedOperation(fmt.Errorf("marshalling checkpoint: %w", err))
+		}
+
+		var fenceUpdate strings.Builder
+		if err := tplUpdateFence.Execute(&fenceUpdate, d.fence); err != nil {
+			return nil, pf.FinishedOperation(fmt.Errorf("evaluating fence template: %w", err))
+		}
+
+		return nil, pf.RunAsyncOperation(func() error {
+			// NB: Motherduck doesn't actually support transactions yet, but this code is written
+			// like it does. Eventually Motherduck will probably support transactions, and the
+			// connector pretending like it already does doesn't hurt anything and may make the
+			// eventual transition easier.
+			txn, err := d.storeConn.BeginTx(ctx, nil)
```

Comment:
Given that motherduck does not support transaction, it means this implementation cannot guarantee exactly-once semantics, right?

Since if `storeQuery` is executed, but `fenceUpdate` isn't, then the transaction might get repeated, and we end up with duplicate data being inserted.

I think given the lack of transactionality, we may want to look into the other pattern for materializations, "Recovery Log is Authoritative with Idempotent Apply":

https://github.com/estuary/flow/blob/f8d02d6211e8208ccbe81331497ff575175f307f/go/protocols/materialize/transactor.go#L58-L77

Although this will require the ability to run the copy from staged file to target table idempotently, which I'm not sure if DuckDB supports. If it doesn't, then the current implementation is the best we can do I guess

Reply:
Yep it will be "at least once".

There's not a direct way to do an idempotent copy/load of files unfortunately. There _may_ be a way to approximate an idempotent load by including the filename of the loaded file(s) as a materialized column (DuckDB makes this available via the `read_json` etc. functions) and running the copy queries in a way that excludes filenames that have already been loaded.

Balancing the effort and potential trickiness (what if the source collection already has a field of `filename`, etc.) of the currently possible `filename`-based idempotent loads vs. the fact that _DuckDB_ does support multi-statement transactions (it's just a MotherDuck limitation) and [MotherDuck lists transactions as a current DuckDB feature that is not yet supported & increased DuckDB feature coverage is continuously being worked on](https://motherduck.com/docs/architecture-and-capabilities/#considerations-and-limitations), I am inclined to wait & hope for transaction support in MotherDuck to resolve this.